src/framework/entity.js

```diff
@@ -587,7 +587,7 @@ class Entity extends GraphNode {
      * Create a deep copy of the Entity. Duplicate the full Entity hierarchy, with all Components
      * and all descendants. Note, this Entity is not in the hierarchy and must be added manually.
      *
-     * @returns {GraphNode} A new Entity which is a deep copy of the original.
+     * @returns {Entity} A new Entity which is a deep copy of the original.
```

Comment:
```suggestion
     * @returns {this} A new Entity which is a deep copy of the original.
```

Reasoning by example:

```js
class UserEntity extends pc.Entity {
  userprop = 123;
  clone() {
    const ret = super.clone();
    ret.userprop = this.userprop;
    return ret;
  }
}
const a = new UserEntity();
const b = a.clone(); // Should be UserEntity and not Entity
```

The `Entity` method `_cloneRecursively(duplicatedIdsMap)` is also wrong, you need to change it to `new this.constructor` instead of `new Entity` (just like in `Material` class).

Reply:
All I'm changing in my pull is the return type to be Entity instead of GraphNode in order that the playcanvas types are correct.

I don't know if you can specify {this} in the typescript type annotations?

Reply:
> I don't know if you can specify {this} in the typescript type annotations?

Sure, I tried to explain it here: https://github.com/playcanvas/engine/pull/4016

As soon as you try to extend from `Entity`, it will fail (as the example shall illustrate), but I know this fixes probably your use case.